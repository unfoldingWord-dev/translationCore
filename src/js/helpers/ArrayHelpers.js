/**
 * @description - Groups consecutive numbers in an array
 * @param {Array} numbers - array of numbers to be grouped
 * @returns {Array} - grouped array of array of consecutive numbers
 */
export const groupConsecutiveNumbers = (numbers) => (
  numbers.reduce(function(accumulator, currentValue, currentIndex, originalArray) {
    if (currentValue) { // ignore undefined entries
      // if this iteration is consecutive to the last, add it to the previous run
      if (currentValue - originalArray[currentIndex - 1] === 1) {
        accumulator[accumulator.length - 1].push(currentValue);
      } else { // the start of a new run including first element
        // create a new subarray with this as the start
        accumulator.push([currentValue]);
      }
    }
    return accumulator; // return state for next iteration
  }, [])
);
/**
 * @description - Deletes indices from an array safely
 * @param {Array} array - array elements to delete from
 * @returns {Array} - the resulting array after indexes were safely removed
 */
export const deleteIndices = (array, indices) => {
  let _array = JSON.parse(JSON.stringify(array));
  indices.sort( (a,b) => b - a );
  indices.forEach(index => {
    if (index >= 0) _array.splice(index, 1);
  });
  return _array;
};
