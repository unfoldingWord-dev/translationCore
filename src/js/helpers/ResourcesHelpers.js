/* eslint-disable no-console */
import fs from 'fs-extra';
import path from 'path-extra';
import ospath from 'ospath';
// helpers
import * as BibleHelpers from './bibleHelpers';
// constants
const USER_RESOURCES_PATH = path.join(ospath.home(), 'translationCore/resources');
const STATIC_RESOURCES_PATH = path.join(__dirname, '../../../tC_resources/resources');

/**
 * Moves all bibles from the static folder to the user's translationCore folder.
 */
export function getBibleFromStaticPackage(force = false) {
  try {
    let languagesIds = ['en', 'grc', 'he', 'hi']; // english, greek, hebrew.
    languagesIds.forEach((languagesId) => {
      const STATIC_RESOURCES_BIBLES_PATH = path.join(STATIC_RESOURCES_PATH, languagesId, 'bibles');
      const BIBLE_RESOURCES_PATH = path.join(USER_RESOURCES_PATH, languagesId, 'bibles');
      let bibleNames = fs.readdirSync(STATIC_RESOURCES_BIBLES_PATH);
      bibleNames.forEach((bibleName) => {
        let bibleSourcePath = path.join(STATIC_RESOURCES_BIBLES_PATH, bibleName);
        let bibleDestinationPath = path.join(BIBLE_RESOURCES_PATH, bibleName);
        if(!fs.existsSync(bibleDestinationPath) || force) {
          fs.copySync(bibleSourcePath, bibleDestinationPath);
        }
      });
    });
  } catch (error) {
    console.error(error);
  }
}

/**
 * @description moves all translationHelps from the static folder to the resources folder in the translationCore folder.
 */
export function getTHelpsFromStaticPackage(force = false) {
  ['en','grc','hi'].forEach(languageId => {
    try {
      const staticTranslationHelpsPath = path.join(STATIC_RESOURCES_PATH, languageId, 'translationHelps');
      const userTranslationHelpsPath = path.join(USER_RESOURCES_PATH, languageId, 'translationHelps');
      const tHelpsNames = fs.readdirSync(staticTranslationHelpsPath);
      tHelpsNames.forEach((tHelpName) => {
        let tHelpSourcePath = path.join(staticTranslationHelpsPath, tHelpName);
        let tHelpDestinationPath = path.join(userTranslationHelpsPath, tHelpName);
        if(!fs.existsSync(tHelpDestinationPath) || force) {
          fs.copySync(tHelpSourcePath, tHelpDestinationPath);
        }
      });
    } catch (error) {
      console.error(error);
    }
  });
}

/**
 * @description moves all translationHelps from the static folder to the resources folder in the translationCore folder.
 */
export function getLexiconsFromStaticPackage(force = false) {
  try {
    const languageId = 'en';
    const staticPath = path.join(STATIC_RESOURCES_PATH, languageId, 'lexicons');
    const userPath = path.join(USER_RESOURCES_PATH, languageId, 'lexicons');
    const folders = fs.readdirSync(staticPath);
    folders.forEach((folder) => {
      let sourcePath = path.join(staticPath, folder);
      let destinationPath = path.join(userPath, folder);
      if(!fs.existsSync(destinationPath) || force) {
        fs.copySync(sourcePath, destinationPath);
      }
    });
  } catch (error) {
    console.error(error);
  }
}

/**
 * @description - Auto generate the chapter index since more projects will use it
 */
export const chapterGroupsIndex = () => {
  const groupsIndex = Array(150).fill().map((_, i) => {
    let chapter = i + 1;
    return {
      id: 'chapter_' + chapter,
      name: 'Chapter ' + chapter
    };
  });
return groupsIndex;
};

export function copyGroupsDataToProjectResources(currentToolName, groupsDataDirectory, bookAbbreviation) {
  const languageId = currentToolName === 'translationWords' ? 'grc' : 'en';
  const toolResourcePath = path.join(USER_RESOURCES_PATH, languageId, 'translationHelps', currentToolName);
  const versionPath = getLatestVersionInPath(toolResourcePath);
  const groupsFolderPath = currentToolName === 'translationWords' ? path.join('kt', 'groups', bookAbbreviation) : path.join('groups', bookAbbreviation);
  const groupsDataSourcePath = path.join(versionPath, groupsFolderPath);

  if(fs.existsSync(groupsDataSourcePath)) {
    fs.copySync(groupsDataSourcePath, groupsDataDirectory);
  } else {
    const groupsData = chapterGroupsData(bookAbbreviation, currentToolName);
    groupsData.forEach(groupData => {
      const groupId = groupData[0].contextId.groupId;
      const chapterIndexPath = path.join(groupsDataDirectory, groupId + '.json');
      fs.outputFileSync(chapterIndexPath, JSON.stringify(groupData, null, 2));
    });
    console.log("Chapter Groups Data generated. translationHelps resources path was not found, " + groupsDataSourcePath);
  }
}
/**
 * @description - Auto generate the chapter index since more projects will use it
 * @param {String} bookId - id of the current book
 * @param {String} currentToolName - id of the current tool
 */
export const chapterGroupsData = (bookId, currentToolName) => {
  let groupsData = [];
  const versionPath = getLatestVersionInPath(path.join(STATIC_RESOURCES_PATH, 'en', 'bibles', 'ulb'));
  const ulbIndexPath = path.join(versionPath, 'index.json');
  if (fs.existsSync(ulbIndexPath)) { // make sure it doens't crash if the path doesn't exist
    const ulbIndex = fs.readJsonSync(ulbIndexPath); // the index of book/chapter/verses
    const bookData = ulbIndex[bookId]; // get the data in the index for the current book
    groupsData = Array(bookData.chapters).fill().map((_, i) => { // create array from number of chapters
      const chapter = i + 1; // index is 0 based, so add one for chapter number
      const verses = bookData[chapter]; // get the number of verses in the chapter
      const groupData = Array(verses).fill().map((_, i) => { // turn number of verses into array
        const verse = i + 1; // index is 0 based, so add one for verse number
        return {
          "contextId": {
            "reference": {
              "bookId": bookId,
              "chapter": chapter,
              "verse": verse
            },
            "tool": currentToolName,
            "groupId": "chapter_" + chapter
          }
        };
      });
      return groupData;
    });
  }
  return groupsData;
};

/**
 * @description Helper function to get a bibles manifest file from the bible resources folder.
 * @param {string} bibleVersionPath - path to a bibles version folder.
 * @param {string} bibleID - bible name. ex. bhp, uhb, udb, ulb.
 */
export function getBibleManifest(bibleVersionPath, bibleID) {
  let fileName = 'manifest.json';
  let bibleManifestPath = path.join(bibleVersionPath, fileName);
  let manifest;

  if(fs.existsSync(bibleManifestPath)) {
    manifest = fs.readJsonSync(bibleManifestPath);
  } else {
    console.error(`Could not find manifest for ${bibleID} at ${bibleManifestPath}`);
  }
  return manifest;
}

/**
 * @description Helper function to get a bibles index from the bible resources folder.
 * @param {string} languageId
 * @param {string} bibleId - bible name. ex. bhp, uhb, udb, ulb.
 * @param {string} bibleVersion - optional release version, if null then get latest
 */
export function getBibleIndex(languageId, bibleId, bibleVersion) {
  const STATIC_RESOURCES_BIBLES_PATH = path.join(__dirname, '../../../tC_resources/resources', languageId, 'bibles');
  const fileName = 'index.json';
  let bibleIndexPath;
  if (bibleVersion) {
    bibleIndexPath = path.join(STATIC_RESOURCES_BIBLES_PATH, bibleId, bibleVersion, fileName);
  } else {
    const versionPath = getLatestVersionInPath(path.join(STATIC_RESOURCES_BIBLES_PATH, bibleId));
    if (versionPath) {
      bibleIndexPath = path.join(versionPath, fileName);
    }
  }
  let index;

  if(fs.existsSync(bibleIndexPath)) {
    index = fs.readJsonSync(bibleIndexPath);
  } else {
    console.error("Could not find index for " + bibleId + ' ' + bibleVersion);
  }
  return index;
}

/**
 * Returns an array of versions found in the path that start with [vV]\d
 * @param {String} resourcePath - base path to search for versions
 * @return {Array} - array of versions, e.g. ['v1', 'v10', 'v1.1']
 */
export function getVersionsInPath(resourcePath) {
  if (! resourcePath || ! fs.pathExistsSync(resourcePath)) {
    return null;
  }
  const isVersionDirectory = name => { 
    const fullPath = path.join(resourcePath, name);
    return fs.lstatSync(fullPath).isDirectory() && name.match(/^v\d/i);
  };
  return fs.readdirSync(resourcePath).filter(isVersionDirectory);
}

/**
 * Returns a sorted an array of versions so that numeric parts are properly ordered (e.g. v10a < v100)
 * @param {Array} - array of versions unsorted: ['v05.5.2', 'v5.5.1', 'V6.21.0', 'v4.22.0', 'v6.1.0', 'v6.1a.0', 'v5.1.0', 'V4.5.0']
 * @return {Array} - array of versions sorted:  ["V4.5.0", "v4.22.0", "v5.1.0", "v5.5.1", "v05.5.2", "v6.1.0", "v6.1a.0", "V6.21.0"]
 */
export function sortVersions(versions) {
  // Don't sort if null, empty or not an array
  if (! versions || ! Array.isArray(versions)) {
    return versions;
  }
  // Only sort of all items are strings
  for(let i = 0; i < versions.length; ++i) {
    if (typeof versions[i] !== 'string') {
      return versions;
    }
  }
  versions.sort( (a, b) => String(a).localeCompare(b, undefined, { numeric:true }) );
  return versions;
}

/**
 * Return the full path to the highest version folder in resource path
 * @param {String} resourcePath - base path to search for versions
 * @return {String} - path to highest version
 */
export function getLatestVersionInPath(resourcePath) {
  const versions = sortVersions(getVersionsInPath(resourcePath));
  if (versions && versions.length) {
    return path.join(resourcePath, versions[versions.length-1]);
  }
  return null; // return illegal path
}

export function getLanguageIdsFromResourceFolder(bookId) {
  let languageIds = fs.readdirSync(USER_RESOURCES_PATH)
    .filter(folder => folder !== '.DS_Store'); // filter out .DS_Store
  // if its an old testament project remove greek from languageIds.
  if (BibleHelpers.isOldTestament(bookId)) {
    languageIds = languageIds.filter(languageId => languageId !== 'grc');
  } else { // else if its a new testament project remove hebrew from languageIds.
    languageIds = languageIds.filter(languageId => languageId !== 'he');
  }
  return languageIds;
}
