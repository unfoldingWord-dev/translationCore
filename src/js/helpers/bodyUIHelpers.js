export function getIconColorFromIndex(stepIndex) {
  let colorsArray = [];
  colorsArray.push(stepIndex === 0 ? "black" : stepIndex > 0 ? "var(--accent-color-dark)" : "");
  colorsArray.push(stepIndex === 1 ? "black" : 1 > stepIndex ? "grey" : stepIndex > 1 ? "var(--accent-color-dark)" : "");
  colorsArray.push(stepIndex === 2 ? "black" : 2 > stepIndex ? "grey" : stepIndex > 2 ? "var(--accent-color-dark)" : "");
  colorsArray.push(stepIndex === 3 ? "black" : 3 > stepIndex ? "grey" : stepIndex > 3 ? "var(--accent-color-dark)" : "");
  return colorsArray;
}