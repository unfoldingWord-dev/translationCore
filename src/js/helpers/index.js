import { generateTimestamp } from './TimestampGenerator.js';
import * as loadHelpers from './LoadHelpers.js';

export {
  generateTimestamp,
  loadHelpers,
};
