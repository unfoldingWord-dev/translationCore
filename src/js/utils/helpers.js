import Path from 'path';
import * as fs from 'fs-extra';
import ManifestGenerator from '../components/core/create_project/ProjectManifest';
import BooksOfBible from '../components/core/BooksOfBible';

module.exports.isOldTestament = function (projectBook) {
    var passedBook = false;
    for (var book in BooksOfBible) {
        if (book == projectBook) passedBook = true;
        if (BooksOfBible[book] == "Malachi" && passedBook) {
            return true;
        }
    }
    return false;
}

module.exports.checkIfValidBetaProject = function (manifest) {
    if (manifest && manifest.project) return manifest.project.id == "eph" || manifest.project.id == "tit";
    else if (manifest && manifest.ts_project) return manifest.ts_project.id == "eph" || manifest.ts_project.id == "tit";
}

/**
 * 
 * @param {*} path 
 * @param {*} file 
 */
export function loadFile(path, file) {
    try {
        var manifest = fs.readJsonSync(Path.join(path, file));
        return manifest;
    }
    catch (e) {
        return null
    }
}

/**
 * @description - Generates and saves a translationCore manifest file
 * @param {string} projectSaveLocation - Filepath of where the translationCore manifest file will
 * be saved. Must be an ABSOLUTE PATH
 * @param {object} data - The translationCore manifest data to be saved
 * @param {object} tsManifest - The translationStudio manifest data loaded from a translation
 * studio project
 */
module.exports.saveManifest = function (projectSaveLocation, link, oldManifest, currentUser) {
    var data = {
        user: [currentUser],
        repo: link
    }
    var manifest;
    try {
        var manifestLocation = Path.join(projectSaveLocation, 'manifest.json');
        if (oldManifest.package_version == '3') {
            //some older versions of ts-manifest have to be tweaked to work
            manifest = this.fixManifestVerThree(oldManifest);
        } else {
            manifest = ManifestGenerator(data, oldManifest);
        }
        fs.outputJsonSync(manifestLocation, manifest);
    }
    catch (err) {
        console.error(err);
    }
    return manifest;
}

/**
 * @desription - Uses the tc-standard format for projects to make package_version 3 compatible
 * @param oldManifest - The name of an employee.
 */
module.exports.fixManifestVerThree = function (oldManifest) {
    var newManifest = {};
    try {
        for (var oldElements in oldManifest) {
            newManifest[oldElements] = oldManifest[oldElements];
        }
        newManifest.finished_chunks = oldManifest.finished_frames;
        newManifest.ts_project = {};
        newManifest.ts_project.id = oldManifest.project_id;
        newManifest.ts_project.name = this.convertToFullBookName(oldManifest.project_id);
        for (var el in oldManifest.source_translations) {
            newManifest.source_translations = oldManifest.source_translations[el];
            var parameters = el.split("-");
            newManifest.source_translations.language_id = parameters[1];
            newManifest.source_translations.resource_id = parameters[2];
            break;
        }
    } catch (e) {
        console.error(e);
    }
    return newManifest;
}

module.exports.convertToFullBookName = function (bookAbbr) {
    if (!bookAbbr) return;
    return BooksOfBible[bookAbbr.toString().toLowerCase()];
}