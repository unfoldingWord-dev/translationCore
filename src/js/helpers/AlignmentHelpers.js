//helpers
import * as VerseObjectHelpers from './VerseObjectHelpers';
import * as ArrayHelpers from './ArrayHelpers';

/**
 * @description pivots alignments into bottomWords/targetLanguage verseObjectArray sorted by verseText
 * @param {Array} alignments - array of aligned word objects {bottomWords, topWords}
 * @param {Array} wordBank - array of topWords
 * @param {String} verseString - The string to base the bottomWords sorting
 * @param {Boolean} useVerseText - if true, then return parsed verse text if unaligned verse has changed, otherwise return null
 * @returns {Array} - sorted array of verseObjects to be used for verseText of targetLanguage
 */
export const merge = (alignments, wordBank, verseString, useVerseText=false) => {
  let verseObjects; // array to return
  // get the definitive list of verseObjects from the verse, unaligned but in order
  const unalignedOrdered = VerseObjectHelpers.getOrderedVerseObjectsFromString(verseString);
  // assign verseObjects with unaligned objects to be replaced with aligned ones
  verseObjects = JSON.parse(JSON.stringify(unalignedOrdered));
  //check each word in the verse string is also in the word bank or alignments
  const verseObjectsNotInAlignmentData = verseStringWordsContainedInAlignments(alignments, wordBank, verseObjects);
  if (verseObjectsNotInAlignmentData.length > 0) {
    if (hasAlignments(alignments)) { // if verse has some alignments
      const verseWordsJoined = verseObjectsNotInAlignmentData.map(({text}) => text).join(', ');
      throw {
        message: `The words "${verseWordsJoined}" from the target language verse are not in the alignment data.`,
        type: 'InvalidatedAlignments'
      };
    } else { // if verse had no alignments
      return useVerseText ? verseObjects : null; // use parsed verse text
    }
  }
  // each wordBank object should result in one verseObject
  for (let bottomWord of wordBank) {
    const verseObject = VerseObjectHelpers.wordVerseObjectFromBottomWord(bottomWord);
    const index = VerseObjectHelpers.indexOfVerseObject(unalignedOrdered, verseObject);
    if (index > -1) {
      verseObjects[index] = verseObject;
    }
    else {
      if (hasAlignments(alignments)) { // if verse has some alignments
        throw {message: `Word "${bottomWord.word}" is in wordBank, but missing from target language verse.`, type: 'InvalidatedAlignments'};
      } else { // if verse had no alignments
        return useVerseText ? verseObjects : null; // use parsed verse text
      }
    }
  }
  let indicesToDelete = [];
  // each alignment should result in one verseObject
  alignments.forEach(alignment => {
    const {topWords, bottomWords} = alignment;
    // each bottomWord results in a nested verseObject of tag: w, type: word
    // located inside innermost nested topWord/k verseObject
    let replacements = {};
    bottomWords.forEach(bottomWord => {
      const verseObject = VerseObjectHelpers.wordVerseObjectFromBottomWord(bottomWord);
      const index = VerseObjectHelpers.indexOfVerseObject(unalignedOrdered, verseObject);
      if (index === -1) {
        throw { message: "VerseObject not found in verseText while merging:" + JSON.stringify(verseObject), type: 'InvalidatedAlignments' };
      }
      replacements[index] = verseObject;
    });
    // each topWord results in a nested verseObject of tag: k, type: milestone
    const milestones = topWords.map(topWord =>
      VerseObjectHelpers.milestoneVerseObjectFromTopWord(topWord)
    );
    const indices = Object.keys(replacements);
    // group consecutive indexes so that they can be aggregated
    const groupedConsecutiveIndices = ArrayHelpers.groupConsecutiveNumbers(indices);
    // loop through groupedConsecutiveIndices to reduce and place where needed.
    groupedConsecutiveIndices.forEach(consecutiveIndices => {
      // map the consecutiveIndices to replacement verseObjects
      const replacementVerseObjects = consecutiveIndices.map(index => replacements[index]);
      // remove and use the first index in group to place the aligned verseObject milestone later
      const indexToReplace = consecutiveIndices.shift();
      // the rest of the consecutiveIndices need to be queued to be deleted later after shift
      indicesToDelete = indicesToDelete.concat(consecutiveIndices);
      // place the replacementVerseObjects in the last milestone as children
      milestones[milestones.length-1].children = replacementVerseObjects;
      // nest the milestones so that the first is the parent and each subsequent is nested
      const milestone = VerseObjectHelpers.nestMilestones(milestones);
      // replace the original verseObject from the verse text with the aligned milestone verseObject
      verseObjects[indexToReplace] = milestone;
    });
  });
  // deleteIndices that were queued due to consecutive bottomWords in alignments
  verseObjects = ArrayHelpers.deleteIndices(verseObjects, indicesToDelete);
  return verseObjects;
};

/**
 * check if there were any alignments
 * @param {Array} alignments
 * @return {boolean} true if an alignment was found
 */
export const hasAlignments = (alignments) => {
  const indexFirstAlignment = alignments.findIndex((alignment) => {
    return alignment.bottomWords.length > 0;
  });
  return indexFirstAlignment >= 0;
};

/**
 * Determines if the given verse objects from a string are contained in
 * the given alignments
 *
 * @param {Array} alignments - array of aligned word objects {bottomWords, topWords}
 * @param {Array} wordBank - array of unused topWords for aligning
 * @param {Object} verseObjects - verse objects from verse string to be checked
 */
export function verseStringWordsContainedInAlignments(alignments, wordBank, verseObjects) {
  return verseObjects.filter((verseObject) => {
    const checkIfWordMatches = function(verseObject) {
      return function({ word, occurrence, occurrences }) {
        const verseObjectWord = verseObject.text;
        const verseObjectOccurrence = verseObject.occurrence;
        const verseObjectOccurrences = verseObject.occurrences;
        return word === verseObjectWord &&
          occurrence === verseObjectOccurrence &&
          occurrences === verseObjectOccurrences;
      };
    };
    if (verseObject.type !== 'word') return false;
    const wordCheckerFunction = checkIfWordMatches(verseObject);
    const containedInWordBank = !!wordBank.find(wordCheckerFunction);
    const containedInAlignments = !!alignments.find(({ bottomWords }) => {
      return !!bottomWords.find(wordCheckerFunction);
    });
    return !containedInWordBank && !containedInAlignments;
  });
}

/**
 * @description pivots alignments into bottomWords/targetLanguage verseObjectArray sorted by verseText
 * @param {Array} verseObjects - array of aligned verseObjects [{milestone children={verseObject}}, ...]
 * @param {Array|Object|String} alignedVerse - optional verse to use for ordering alignments
 * @returns {Object} - object of alignments (array of alignments) and wordbank (array of unused words)
 */
export const unmerge = (verseObjects, alignedVerse) => {
  let baseMilestones = [], wordBank = [];
  let alignments = [];
  if (verseObjects && verseObjects.verseObjects) {
    verseObjects = verseObjects.verseObjects;
  }
  if (typeof alignedVerse !== 'string') {
    alignedVerse = VerseObjectHelpers.getWordList(alignedVerse);
  }
  for (let verseObject of verseObjects) {
    let alignment = getAlignmentForMilestone(baseMilestones, verseObject);
    if (!alignment) {
      alignment = {topWords: [], bottomWords: []};
      alignments.push(alignment);
      baseMilestones.push({alignment: alignment, milestone: verseObject});
    }
    addVerseObjectToAlignment(verseObject, alignment);
  }
  const alignmentUnOrdered = [];
  for (let _alignment of alignments) {
    if (_alignment.topWords.length > 0) {
      alignmentUnOrdered.push(_alignment);
    } else {
      wordBank = wordBank.concat(_alignment.bottomWords);
    }
  }
  let alignment = orderAlignments(alignedVerse, alignmentUnOrdered);
  return { alignment, wordBank};
};

/**
 * @description uses the alignedVerseString to order alignments
 * @param {String|Array} alignmentVerse - optional alignment verse
 * @param {Array} alignmentUnOrdered - alignments to order
 * @return {Array} ordered alignments if alignment string given, else unordered alignments
 */
export const orderAlignments = function (alignmentVerse, alignmentUnOrdered) {
  if (typeof alignmentVerse === 'string') {
    alignmentVerse = VerseObjectHelpers.getOrderedVerseObjectsFromString(alignmentVerse);
  } else {
    alignmentVerse = VerseObjectHelpers.getOrderedVerseObjects(alignmentVerse);
  }
  if (Array.isArray(alignmentVerse)) {
    let alignment = [];
    // order alignments
    for (let i = 0; i < alignmentVerse.length; i++) {
      const nextWord = alignmentVerse[i];
      let index = indexOfFirstMilestone(alignmentUnOrdered, nextWord);
      if ((index < 0) && (nextWord.type === 'word') && (i < alignmentVerse.length - 1)) {
        const wordAfter = alignmentVerse[i + 1];
        if (wordAfter.type === 'text') { // maybe this was punctuation split from word
          nextWord.text += wordAfter.text; // add possible punctuation
          index = indexOfFirstMilestone(alignmentUnOrdered, nextWord); // try again
        }
      }
      if (index >= 0) {
        alignment.push(alignmentUnOrdered[index]);
        alignmentUnOrdered.splice(index, 1); // remove item
      } else { // if not found, may be either an unaligned topWord or merged topWord
        if (nextWord.type === 'word') {
          index = indexOfMilestone(alignmentUnOrdered, nextWord);
          if (index < 0) { // if not found in unordered list, try already ordered
            index = indexOfMilestone(alignment, nextWord);
          }
          if (index < 0) { // if still not found in topWords, it's an unaligned topWord
            const wordObject = VerseObjectHelpers.alignmentObjectFromVerseObject(nextWord);
            alignment.push({topWords: [wordObject], bottomWords: []});
          }
        }
      }
    }
    if (alignmentUnOrdered.length > 0) {
      alignment = alignment.concat(alignmentUnOrdered);
    }
    return alignment;
  }
  return alignmentUnOrdered;
};

/**
 * compare occurrences of a and b, and handle conversion to int if necessary
 * @param a
 * @param b
 * @return {boolean}
 */
const compareOccurrences = function (a, b) {
  let sameOccurrence = (a.occurrence === b.occurrence);
  if (!sameOccurrence && a.occurrence && b.occurrence) {
    if (typeof a.occurrence !== typeof b.occurrence) { // one may be string and the other an int
      const occurrence1 = (typeof a.occurrence === 'string') ? parseInt(a.occurrence) : a.occurrence;
      const occurrence2 = (typeof b.occurrence === 'string') ? parseInt(b.occurrence) : b.occurrence;
      sameOccurrence = (occurrence1 === occurrence2) && (occurrence1 !== 0);
    }
  }
  return sameOccurrence;
};
/**
 * @description Returns index of the verseObject in the alignments first milestone (ignores occurrences since that can be off)
 * @param {Array} alignments - array of the alignments to search in
 * @param {Object} verseObject - verseObject to search for
 * @returns {Int} - the index of the verseObject
 */
export const indexOfFirstMilestone = (alignments, verseObject) => {
  let index = -1;
  if (verseObject.type === 'word') {
    index = alignments.findIndex(alignment => {
      if (alignment.topWords.length > 0) {
        const _verseObject = alignment.topWords[0];
        if (_verseObject.word === verseObject.text) {
          return compareOccurrences(_verseObject, verseObject);
        }
      }
      return false;
    });
  }
  return index;
};
/**
 * @description Returns index of the verseObject in the alignments milestone (ignores occurrences since that can be off)
 * @param {Array} alignments - array of the alignments to search in
 * @param {Object} verseObject - verseObject to search for
 * @returns {Int} - the index of the verseObject
 */
export const indexOfMilestone = (alignments, verseObject) => {
  let index = -1;
  if (verseObject.type === 'word') {
    index = alignments.findIndex(alignment => {
      for (let _verseObject of alignment.topWords) {
        if (_verseObject.word === verseObject.text) {
          return compareOccurrences(_verseObject, verseObject);
        }
      }
      return false;
    });
  }
  return index;
};

/**
 * @description find the alignment to use for this milestone.  If milestone has already been given an alignment, then
 *                use that one.  Otherwise return null.  This is needed because milestones are not always
 *                contiguous.
 * @param {Array} baseMilestones - already found base milestones.
 * @param {Object} newMilestone
 * @return {Object} previous Alignment if found - else null.
 */
const getAlignmentForMilestone = (baseMilestones, newMilestone) => {
  for (let baseMilestone of baseMilestones) {
    if (baseMilestone.alignment && sameMilestone(baseMilestone.milestone, newMilestone)) {
      return baseMilestone.alignment;
    }
  }
  return null;
};

/**
 * @description test to see if this is the same milestone (needed when milestones are not contiguous)
 * @param {Object} a
 * @param {Object} b
 * @return {boolean} true if same milestone
 */
const sameMilestone = (a, b) => {
  const same = (a.type === b.type) && (a.content === b.content) && (a.occurrence === b.occurrence);
  return same;
};

/**
 * @description adds verse object to alignment
 * @param {Object} verseObject
 * @param {Object} alignment
 */
export const addVerseObjectToAlignment = (verseObject, alignment) => {
  if (verseObject.type === 'milestone' && verseObject.children.length > 0) {
    const wordObject = VerseObjectHelpers.alignmentObjectFromVerseObject(verseObject);
    const duplicate = alignment.topWords.find(function (obj) {
      return (obj.word === wordObject.word) && (obj.occurrence === wordObject.occurrence) ;
    });
    if (!duplicate) {
      alignment.topWords.push(wordObject);
    }
    verseObject.children.forEach(_verseObject => {
      addVerseObjectToAlignment(_verseObject, alignment);
    });
  } else if (verseObject.type === 'word' && !verseObject.children) {
    const wordObject = VerseObjectHelpers.alignmentObjectFromVerseObject(verseObject);
    alignment.bottomWords.push(wordObject);
  }
};

