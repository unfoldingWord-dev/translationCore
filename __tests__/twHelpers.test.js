import fs from 'fs-extra';
import path from 'path-extra';
import ospath from 'ospath';
import * as twHelpers from '../src/js/helpers/twHelpers';
import * as ResourcesHelpers from '../src/js/helpers/ResourcesHelpers';

describe('Test tW Helpers', function() {
  const lang = 'grc';
  const bible = 'ugnt';
  const ugntRealPath = path.join(__dirname, 'fixtures/resources/grc/bibles/ugnt');
  const ugntMockPath = path.join(ospath.home(), 'translationCore/resources', lang, 'bibles', bible);
  const twBasePath = path.join(ospath.home(), 'translationCore/resources', lang, 'translationHelps', 'translationWords');

  beforeEach(() => {
    fs.__resetMockFS();
    fs.__loadDirIntoMockFs(ugntRealPath, ugntMockPath);
  });

  afterEach(() => {
    fs.__resetMockFS();
  });

  it('Test that milestones are properly constructed using inchrist for phm', () => {
    // given
    const expectedTwPath = path.join(twBasePath, path.basename(ResourcesHelpers.getLatestVersionInPath(ugntMockPath)));

    // when
    const generatedTwPath = twHelpers.generateTw(lang, bible);
    const jsonFile = path.join(generatedTwPath, 'kt', 'groups', 'phm', 'inchrist.json');

    // then
    expect(generatedTwPath).toEqual(expectedTwPath);
    expect(fs.existsSync(jsonFile)).toBeTruthy();
    const data = JSON.parse(fs.readFileSync(jsonFile));
    expect(data).toMatchSnapshot();
    const expectedItems = 5;
    expect(data.length).toEqual(expectedItems);
  });

  it('Test that occurrence of God is correct in Titus 1:1', () => {
    // given
    const expectedTwPath = path.join(twBasePath, path.basename(ResourcesHelpers.getLatestVersionInPath(ugntMockPath)));

    // when
    const generatedTwPath = twHelpers.generateTw(lang, bible);


    // then
    expect(generatedTwPath).toEqual(expectedTwPath);
    const jsonFile = path.join(generatedTwPath, 'kt', 'groups', 'tit', 'god.json');
    expect(fs.existsSync(jsonFile)).toBeTruthy();
    const data = JSON.parse(fs.readFileSync(jsonFile));
    const expectedOccurrence = 2;
    expect(data[1].contextId.occurrence).toEqual(expectedOccurrence);
  });
});
