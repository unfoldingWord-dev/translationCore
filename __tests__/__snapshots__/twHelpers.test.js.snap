// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test tW Helpers Test that milestones are properly constructed using inchrist for phm 1`] = `
Array [
  Object {
    "comments": false,
    "contextId": Object {
      "groupId": "inchrist",
      "occurrence": 1,
      "quote": "ἐν Χριστῷ",
      "reference": Object {
        "bookId": "phm",
        "chapter": 1,
        "verse": 8,
      },
      "strong": Array [
        "G17220",
        "G55470",
      ],
      "tool": "translationWords",
    },
    "priority": 1,
    "reminders": false,
    "selections": false,
    "verseEdits": false,
  },
  Object {
    "comments": false,
    "contextId": Object {
      "groupId": "inchrist",
      "occurrence": 1,
      "quote": "ἐν Κυρίῳ",
      "reference": Object {
        "bookId": "phm",
        "chapter": 1,
        "verse": 16,
      },
      "strong": Array [
        "G17220",
        "G29620",
      ],
      "tool": "translationWords",
    },
    "priority": 1,
    "reminders": false,
    "selections": false,
    "verseEdits": false,
  },
  Object {
    "comments": false,
    "contextId": Object {
      "groupId": "inchrist",
      "occurrence": 1,
      "quote": "ἐν Κυρίῳ",
      "reference": Object {
        "bookId": "phm",
        "chapter": 1,
        "verse": 20,
      },
      "strong": Array [
        "G17220",
        "G29620",
      ],
      "tool": "translationWords",
    },
    "priority": 1,
    "reminders": false,
    "selections": false,
    "verseEdits": false,
  },
  Object {
    "comments": false,
    "contextId": Object {
      "groupId": "inchrist",
      "occurrence": 1,
      "quote": "ἐν Χριστῷ",
      "reference": Object {
        "bookId": "phm",
        "chapter": 1,
        "verse": 20,
      },
      "strong": Array [
        "G17220",
        "G55470",
      ],
      "tool": "translationWords",
    },
    "priority": 1,
    "reminders": false,
    "selections": false,
    "verseEdits": false,
  },
  Object {
    "comments": false,
    "contextId": Object {
      "groupId": "inchrist",
      "occurrence": 1,
      "quote": "ἐν Χριστῷ Ἰησοῦ",
      "reference": Object {
        "bookId": "phm",
        "chapter": 1,
        "verse": 23,
      },
      "strong": Array [
        "G17220",
        "G55470",
        "G24240",
      ],
      "tool": "translationWords",
    },
    "priority": 1,
    "reminders": false,
    "selections": false,
    "verseEdits": false,
  },
]
`;
